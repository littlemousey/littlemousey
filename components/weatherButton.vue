<template>
  <v-btn v-if="weatherData" text class="font-weight-bold" nuxt to="weather"
    ><span class="weather-icon">{{ weatherIcon }}</span
    >{{ weatherData.temperature }}°C<span class="weather-icon">{{
      dayNightIndicator
    }}</span>
  </v-btn>
</template>

<script>
import { mapActions, mapState } from 'vuex'
export default {
  name: 'WeatherButton',
  data() {
    return {
      weatherData: null,
      weatherIcons: [
        { name: 'wind', icon: '💨' },
        { name: 'bewolkt', icon: '☁️' },
        { name: 'zonnig', icon: '☀️' },
        { name: 'regen', icon: '🌧️' },
        { name: 'onweer', icon: '🌩️' },
        { name: 'sneeuw', icon: '🌨️' },
        { name: 'vorst', icon: '❄️' },
        { name: 'overig', icon: '🌤️' }
      ]
    }
  },
  computed: {
    weatherIcon() {
      const currentWeatherDisplay = this.weatherIcons.find(
        (weatherOption, index, optionList) => {
          return this.weatherData.weatherdescription.includes(
            weatherOption.name
          )
        }
      )
      return currentWeatherDisplay
        ? currentWeatherDisplay.icon
        : this.weatherIcons[this.weatherIcons.length - 1].icon
    },
    ...mapState('weather', ['sunrise', 'sunset']),
    dayNightIndicator() {
      const dutchTime = parseInt(this.weatherData.timestamp.substring(11, 13))
      const sunrise = parseInt(this.sunrise.substring(11, 13))
      const sunset = parseInt(this.sunset.substring(11, 13))
      if (dutchTime < sunrise || dutchTime > sunset) {
        return '🌙'
      }
      return '☀️'
    }
  },
  mounted() {
    this.checkWeatherData()
    const self = this
    setInterval(function() {
      self.checkWeatherData()
    }, 600000)
  },
  methods: {
    async checkWeatherData() {
      const data = await this.$axios.$get(
        'https://data.buienradar.nl/2.0/feed/json'
      )
      const nationalWeather = data.actual.stationmeasurements.find(station => {
        return station.stationname === 'Meetstation De Bilt'
      })
      this.weatherData = nationalWeather
      this.setWeatherData(this.weatherData)
      const sunData = {
        sunrise: data.actual.sunrise,
        sunset: data.actual.sunset
      }

      this.setSunsetData(sunData)
    },
    ...mapActions('weather', ['setWeatherData']),
    ...mapActions('weather', ['setSunsetData'])
  }
}
</script>

<style lang="scss" scoped>
.weather-icon {
  font-size: 25px;
  margin: 0 5px;
}
</style>
