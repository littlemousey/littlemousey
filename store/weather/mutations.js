export default {
  setWeatherData(state, payload) {
    state.weatherData = payload
  },
  setSunsetData(state, { sunrise, sunset }) {
    state.sunrise = sunrise
    state.sunset = sunset
  }
}
