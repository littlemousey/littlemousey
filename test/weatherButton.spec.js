import { mount } from '@vue/test-utils'
import weatherButton from '@/components/weatherButton'
import mutations from '@/store/weather/mutations.js'

let wrapper, vm
const mockFunction = jest.fn().mockImplementation(() => Promise.resolve(true))

beforeEach(() => {
  wrapper = mount(weatherButton, {
    methods: {
      checkWeatherData: mockFunction
    }
  })
  vm = wrapper.vm
  vm.checkWeatherData = mockFunction
})

describe('WeatherButton', () => {
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('state about sunset and sunrise should be changed after retrieving weatherdata', () => {
    const sunData = {
      sunrise: '2019-08-06T06:08:00',
      sunset: '2019-08-06T21:22:00'
    }
    const state = {
      sunrise: null,
      sunset: null
    }
    mutations.setSunsetData(state, sunData)

    expect(state).toEqual({
      sunrise: '2019-08-06T06:08:00',
      sunset: '2019-08-06T21:22:00'
    })
  })
})
