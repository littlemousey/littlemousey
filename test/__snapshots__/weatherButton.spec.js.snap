// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherButton renders correctly 1`] = `<!---->`;
